source/core.py

```diff
@@ -251,6 +251,54 @@ def onEndSession(evt):
 	import windowUtils
 	class MessageWindow(windowUtils.CustomWindow):
 		className = u"wxWindowClassNR"
+		oldBatteryStatus = None
+		hasBeenInitialized = False
+		#Just define these constants here, so we don't have to import win32con
+		WM_POWERBROADCAST = 0x218
+		WM_DISPLAYCHANGE = 0x7e
+		PBT_APMPOWERSTATUSCHANGE = 0xA
+		UNKNOWN_BATTERY_STATUS = 0xFF
+		AC_ONLINE = 0X1
+		NO_SYSTEM_BATTERY = 0X80
+
+		def __init__(self, windowName=None):
+			super(MessageWindow, self).__init__(windowName)
+			self.handlePowerStatusChange()
+
+		def windowProc(self, hwnd, msg, wParam, lParam):
+			import ui
+			if msg == self.WM_POWERBROADCAST and wParam == self.PBT_APMPOWERSTATUSCHANGE:
+				self.handlePowerStatusChange()
+			# Resolution detection comes from an article found at https://msdn.microsoft.com/en-us/library/ms812142.aspx.
+			elif msg == self.WM_DISPLAYCHANGE:
+				if lParam%0x10000 > lParam/0x10000:
+					#Translators: The screen is oriented so that it is wider than it is tall.
+					ui.message(_("Landscape" ))
+				else:
+					#Translators: The screen is oriented in such a way that the height is taller than it is wide.
+					ui.message(_("Portrait"))
+
+		def handlePowerStatusChange(self):
+			#Mostly taken from script_say_battery_status, but modified.
+			import ui
+			import winKernel
+			sps = winKernel.SYSTEM_POWER_STATUS()
+			if not winKernel.GetSystemPowerStatus(sps) or sps.BatteryFlag is self.UNKNOWN_BATTERY_STATUS:
+				return
+			if sps.BatteryFlag & self.NO_SYSTEM_BATTERY or sps.ACLineStatus == self.oldBatteryStatus:
+				return
+			#Necessary because sometimes this event double fires, and  if the battery decreases by 3 percent, this event occurs.
+			MessageWindow.oldBatteryStatus = sps.ACLineStatus
+			if MessageWindow.hasBeenInitialized == False:
+				MessageWindow.hasBeenInitialized = True
+				return #We don't actually want to output anything, just initialize the cache.

```

Comment:
Haha, I was working on lots of c++ at the time, so my brain was in type safety mode.